crates/bindings/src/table.rs

```diff
@@ -0,0 +1,414 @@
+    ///
+    /// Implementors are encouraged to include the table name, unique column name, and unique column value
+    /// in the panic message when no such row previously existed.
+    fn update(&self, new_row: Self::Row) -> Self::Row;
+}
+
+pub trait BTreeIndex {
+    type Row;
+    type IndexType;
+
+    /// Returns an iterator over all rows in the database state where the indexed column(s) match the bounds `b`.
+    ///
+    /// `b` may be:
+    /// - A value for the first indexed column.
+    /// - A range of values for the first indexed column.
+    /// - A tuple of values for any prefix of the indexed columns, optionally terminated by a range for the next.
+    fn filter<B: BTreeIndexBounds<Self::IndexType, K>, K>(&self, b: B) -> impl Iterator<Item = Self::Row>;
+
+    /// Deletes all rows in the database state where the indexed column(s) match the bounds `b`.
+    ///
+    /// `b` may be:
+    /// - A value for the first indexed column.
+    /// - A range of values for the first indexed column.
+    /// - A tuple of values for any prefix of the indexed columns, optionally terminated by a range for the next.
+    ///
+    /// May panic if deleting any one of the rows would violate a constraint,
+    /// though as of proposing no such constraints exist.
+    fn delete<B: BTreeIndexBounds<Self::IndexType, K>, K>(&self, b: B) -> u64;
+}
+
+pub trait BTreeIndexBounds<T, K = ()> {}
+
+macro_rules! impl_btree_index_bounds {
+    ($T:ident $(, $U:ident)*) => {
+        impl_btree_index_bounds!(@impl (), ($T $(, $U)*));
+
+        impl_btree_index_bounds!($($U),*);
+    };
+    () => {};
+    (@impl ($($V:ident),*), ($T:ident $(, $U:ident)+)) => {
+        impl<$($V,)* $T, $($U,)+ Term: BTreeIndexBoundsTerminator<$T>> BTreeIndexBounds<($($U,)+ $T, $($V,)*)> for ($($U,)+ Term,) {}
+        impl_btree_index_bounds!(@impl ($($V,)* $T), ($($U),*));
+    };
+    (@impl ($($V:ident),*), ($T:ident)) => {
+        impl<$($V,)* $T, Term: BTreeIndexBoundsTerminator<$T>> BTreeIndexBounds<($T, $($V,)*)> for (Term,) {}
+        impl<$($V,)* $T, Term: BTreeIndexBoundsTerminator<$T>> BTreeIndexBounds<($T, $($V,)*), SingleBound> for Term {}
+    };
+}
+
+pub struct SingleBound;
+
+impl_btree_index_bounds!(A, B, C, D, E, F);
+
+pub trait BTreeIndexBoundsTerminator<T> {}
+
+impl<T> BTreeIndexBoundsTerminator<T> for T {}
+impl<T> BTreeIndexBoundsTerminator<T> for &T {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::Range<T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeFrom<&T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeFrom<T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeInclusive<&T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeInclusive<T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeTo<&T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeTo<T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeToInclusive<&T> {}
+impl<T> BTreeIndexBoundsTerminator<T> for ops::RangeToInclusive<T> {}
```

Comment:
We need to discuss (on the proposal) whether we actually want to accept all combinations of `T` and `&T`.